design/design-doc-webhooks.md

```diff
@@ -0,0 +1,383 @@
+webhook:
+```golang
+// Convertible defines capability of a type to convertible i.e. it can be converted to/from a hub type.
+type Convertible interface {
+  runtime.Object
+  ConvertTo(dst Hub) error
+  ConvertFrom(src Hub) error
+}
+
+// Hub marks that a given type is the hub type for conversion. This means that
+// all conversions will first convert to the hub type, then convert from the hub
+// type to the destination type. All types besides the hub type should implement
+// Convertible.
+type Hub interface {
+  runtime.Object
+  Hub()
+}
+```
+
+We see that the interfaces include `runtime.Object`, which means the functions
+need to be members of the kinds, i.e. `RDSInstance` struct needs to implements
+these functions. The abstraction layer that accepts `Defaulter` and `Validator`
+look pretty similar so I'll copy only the [mutation
+one](https://github.com/kubernetes-sigs/controller-runtime/blob/0208f43/pkg/webhook/admission/defaulter.go#L54):
+
+```golang
+// Handle handles admission requests.
+func (h *mutatingHandler) Handle(ctx context.Context, req Request) Response {
+  if h.defaulter == nil {
+    panic("defaulter should never be nil")
+  }
+
+  // Get the object in the request
+  obj := h.defaulter.DeepCopyObject().(Defaulter)
+  err := h.decoder.Decode(req, obj)
+  if err != nil {
+    return Errored(http.StatusBadRequest, err)
+  }
+
+  // Default the object
+  obj.Default()
+  marshalled, err := json.Marshal(obj)
+  if err != nil {
+    return Errored(http.StatusInternalServerError, err)
+  }
+
+  // Create the patch
+  return PatchResponseFromRaw(req.Object.Raw, marshalled)
+}
+```
+
+The conversion webhook implementation is a bit [more
+complex](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/webhook/conversion/conversion.go#L22)
+due to the marshalling and version checks that need to happen. However, it seems
+that some of the cases it needs to handle [are not
+necessary](https://github.com/kubernetes-sigs/controller-runtime/issues/1388) in
+v1 CRD mechanics.
+
+These high level abstractions are fairly simple but they have a few drawbacks:
+* They require implementation in API type packages and we want to keep API type
+  packages as lean as possible so that external users of those Go packages do
+  not have to depend on things that are irrelevant to them. However, there are
+  some cases that might require including cloud provider SDK and/or `client`
+  package from controller-runtime:
+  * Let's say GCP decides to add a new field in the new version of SDK and we
+    want to include it in the new `apiVersion`. However, in order to make
+    queries the value of that field needs to be known, hence we need to make a
+    `GET` call to the provider API, which means we'd have to import GCP Go SDK.
+    We update SDKs very infrequently so we have not encountered this but it's
+    possible.
```

Comment:
I'm having a tough time wrapping my head around this case. Could you elaborate on it a little?

Reply:
Yes, for example, let's say we're migrating from a Jet CRD to native or vice versa, and the resource definitions do not match, i.e. `RouteTable` has `ingress` array field in one but not in the other because it has `RouteTableIngress`. In such cases, we'd need to make calls to AWS to get the information about the ingresses so that the conversion is loss-less.

Reply:
Though that's adressed with custom handlers now https://github.com/crossplane/crossplane/pull/2719/files#r817234614

Reply:
I've removed this section completely, proposing that we build layer only to have an ordered execution of multiple functions in a given webhook call; no re-implementation since untyped webhook registration is now possible.